File: Algorithms/1448_Count_Good_Nodes_in_Binary_Tree/Java/CountGoodNodesinBinaryTreeSolution2.java

// Space: O(1)
// Time: O(n)

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int count = 0;
    public int goodNodes(TreeNode root) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return 1;
        dfs(root,root.val);
        return count;
    }

    private void dfs(TreeNode root, int point) {
        if (root == null) return;
        if (root.val >= point) count++;
        if (root.left != null) dfs(root.left, Math.max(root.val, point));
        if (root.right != null) dfs(root.right, Math.max(root.val, point));
    }
}

File: Algorithms/1448_Count_Good_Nodes_in_Binary_Tree/Python/Count_Good_Nodes_in_Binary_Tree_Solution_2.py

# Space: O(1)
# Time: O(n)

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root):
        if root is None: return 0
        if root.left is None and root.right is None: return 1
        self.count = 0

        def dfs(root, point):
            if root is None: return
            if root.val >= point: self.count += 1

            if root.left:
                dfs(root.left, max(point, root.val))
            if root.right:
                dfs(root.right, max(point, root.val))

        dfs(root,root.val)
        return self.count




